src/main/java/ch/njol/skript/registrations/EventValues.java

```diff
@@ -271,58 +271,79 @@ public T get(E event) {
 				}
 			};
 		}
-		// Fourth check will attempt to convert the event value to the type.
+		// Fourth check will attempt to convert the event value to the requesting type.
+		// This first for loop will check that the events are exact. See issue #5016
 		for (EventValueInfo<?, ?> ev : eventValues) {
-			boolean checkInstanceOf = !ev.event.isAssignableFrom(e);
-			if (checkInstanceOf && !e.isAssignableFrom(ev.event))
+			if (!event.isInstance(ev.event))
```

Comment:
`event.isInstance(ev.event)` is always false: `event is a Class<? extends Event>` so isInstance can only be true if called with an event instance. `ev.event` is a class.